lib/srv/db/common/reporting.go

```diff
@@ -0,0 +1,272 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"net"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/observability/metrics"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+type reporterConfig struct {
+	engine    Engine
+	clock     clockwork.Clock
+	component string
+	database  types.Database
+}
+
+func (r *reporterConfig) CheckAndSetDefaults() error {
+	if r.engine == nil {
+		return trace.BadParameter("missing parameter engine")
+	}
+	if r.database == nil {
+		return trace.BadParameter("missing parameter database")
+	}
+	if r.clock == nil {
+		r.clock = clockwork.NewRealClock()
+	}
+	if r.component == "" {
+		r.component = teleport.ComponentDatabase
+	}
+	return nil
+}
+
+type reportingEngine struct {
+	reporterConfig
+}
+
+func init() {
+	_ = metrics.RegisterPrometheusCollectors(prometheusCollectorsEngine...)
+}
+
+func NewReportingEngine(cfg reporterConfig) (Engine, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &reportingEngine{cfg}, nil
+}
+
+func (r *reportingEngine) InitializeConnection(clientConn net.Conn, sessionCtx *Session) error {
```

Comment:
I would suggest to drop this metrics for now or add this on the teleport proxy side instead of teleport db agent side to have more insight in Teleport Cloud.

Reply:
Ok, dropped.